### PR TITLE
Add formatErrorMsg for comparing error messages in tests

### PR DESCRIPTION
**Pull request by lilgreenbird**
formatErrorMsg() from @rene-ye

### CODE REVIEW

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 1]
```diff
@@ -2032,13 +1994,9 @@ private void login(String primary, String primaryInstanceName, int primaryPortNu
...
         // For non-dbmirroring, non-tnir and non-multisubnetfailover scenarios, full time out would be used as time
         // slice.
-        if (isDBMirroring || useParallel) {
-            timeoutUnitInterval = (long) (TIMEOUTSTEP * timerTimeout);
-        } else if (useTnir) {
-            timeoutUnitInterval = (long) (TIMEOUTSTEP_TNIR * timerTimeout);
-        } else {
-            timeoutUnitInterval = timerTimeout;
-        }
+        timeoutUnitInterval = isDBMirroring || useParallel ? (long) (TIMEOUTSTEP * timerTimeout)
```
Comment (cheenamalhotra) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]:
```suggestion
        timeoutUnitInterval = (isDBMirroring || useParallel) ? (long) (TIMEOUTSTEP * timerTimeout)
```

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 2]
```diff
@@ -1915,11 +1882,8 @@ else if (0 == requestedPacketSize)
 
             // transparentNetworkIPResolution is ignored if multiSubnetFailover or DBMirroring is true and user does not
             // set TNIR explicitly
-            if (multiSubnetFailover || (null != failOverPartnerPropertyValue)) {
-                if (!userSetTNIR) {
-                    transparentNetworkIPResolution = false;
-                }
-            }
+            transparentNetworkIPResolution = !((multiSubnetFailover || (null != failOverPartnerPropertyValue))
```
Comment (cheenamalhotra) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: Do we want to set to `true` for alternative case?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 3]
```diff
@@ -1684,12 +1660,9 @@ Connection connectInternal(Properties propsIn,
             }
 
             // Turn off TNIR for FedAuth if user does not set TNIR explicitly
-            if (!userSetTNIR) {
-                if ((!authenticationString.equalsIgnoreCase(SqlAuthentication.NotSpecified.toString()))
-                        || (null != accessTokenInByte)) {
-                    transparentNetworkIPResolution = false;
-                }
-            }
+            transparentNetworkIPResolution = userSetTNIR
```
Comment (cheenamalhotra) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: same comment here, do we want to set to true ?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 4]
```diff
@@ -1119,25 +1106,28 @@ protected boolean needsReconnect() throws SQLServerException {
...
      */
-    private boolean booleanPropertyOn(String propName, String propValue) throws SQLServerException {
+    private boolean isBooleanPropertyOn(String propName, String propValue) throws SQLServerException {
         // Null means take the default of false.
         if (null == propValue)
             return false;
-        String lcpropValue = propValue.toLowerCase(Locale.US);
 
-        if ("true".equals(lcpropValue))
+        String lcpropValue = propValue.toLowerCase(Locale.US);
```
Comment (cheenamalhotra) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: instead of converting to lowercase, compare with equalsIgnoreCase

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 5]
```diff
@@ -4823,48 +4722,42 @@ final boolean complete(LogonCommand logonCommand, TDSReader tdsReader) throws SQ
...
+        // Denali --> TDS 7.4, Katmai (10.0) & later 7.3B, Prelogin disconnects anything older
+        tdsVersion = serverMajorVersion >= 11 ? TDS.VER_DENALI
+                                              : serverMajorVersion >= 10 ? TDS.VER_KATMAI
+                                                                         : serverMajorVersion >= 9 ? TDS.VER_YUKON
+                                                                                                   : TDS.VER_UNKNOWN;
+        if (tdsVersion == TDS.VER_UNKNOWN) {
             assert false : "prelogin did not disconnect for the old version: " + serverMajorVersion;
         }
 
+        final int TDS_LOGIN_REQUEST_BASE_LEN = 94;
```
[the remark this answers is not in the document]
Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: it's not an add, it's just moving a line from above to lower where it's used

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 6]
```diff
@@ -1184,9 +1172,7 @@ Connection connect(Properties propsIn, SQLServerPooledConnection pooledConnectio
                         String sPropValue = propsIn.getProperty(SQLServerDriverIntProperty.LOGIN_TIMEOUT.toString());
                         if (null != sPropValue && sPropValue.length() > 0) {
                             int sPropValueInt = Integer.parseInt(sPropValue);
-                            if (0 != sPropValueInt) { // Use the default timeout in case of a zero value
-                                loginTimeoutSeconds = sPropValueInt;
-                            }
+                            loginTimeoutSeconds = 0 != sPropValueInt ? sPropValueInt : loginTimeoutSeconds;
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: revert this, possible unnecessary self-assignment

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 7]
```diff
@@ -1344,16 +1330,14 @@ Connection connectInternal(Properties propsIn,
...
 
             // get the server name from the properties if it has instance name in it, getProperty the instance name
             // if there is a port number specified do not get the port number from the instance name
             sPropKey = SQLServerDriverStringProperty.SERVER_NAME.toString();
             sPropValue = activeConnectionProperties.getProperty(sPropKey);
 
-            if (sPropValue == null) {
-                sPropValue = "localhost";
-            }
+            sPropValue = null == sPropValue ? "localhost" : sPropValue;
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: same here as above

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 8]
```diff
@@ -2169,6 +2115,7 @@ private void login(String primary, String primaryInstanceName, int primaryPortNu
                 if ((SQLServerException.LOGON_FAILED == sqlex.getErrorCode()) // actual logon failed, i.e. bad password
                         || (SQLServerException.PASSWORD_EXPIRED == sqlex.getErrorCode()) // actual logon failed, i.e.
                                                                                          // password isExpired
+                                                                                         // password isExpired
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: is this comment line duplicated for some reason?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 9]
```diff
@@ -280,12 +281,7 @@ private boolean isDiscarded() {
          *         associated with it. true: Reference was successfully added.
          */
         boolean tryAddReference() {
-            if (isDiscarded() || isExplicitlyDiscarded())
-                return false;
-            else {
-                int refCount = handleRefCount.incrementAndGet();
-                return refCount > 0;
-            }
+            return isDiscarded() || isExplicitlyDiscarded() ? false : handleRefCount.incrementAndGet() > 0;
```
Comment (cheenamalhotra) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: I think I mentioned similar comments before.. please make sure conditions are always in brackets if there's OR or AND clause in the condition - so the interpreter does not get confused.
```suggestion
            return (isDiscarded() || isExplicitlyDiscarded()) ? false : handleRefCount.incrementAndGet() > 0;
```

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 10]
```diff
@@ -709,18 +703,13 @@ public static synchronized void registerColumnEncryptionKeyStoreProviders(
 
     static synchronized SQLServerColumnEncryptionKeyStoreProvider getGlobalSystemColumnEncryptionKeyStoreProvider(
             String providerName) {
-        if (null != globalSystemColumnEncryptionKeyStoreProviders
-                && globalSystemColumnEncryptionKeyStoreProviders.containsKey(providerName)) {
-            return globalSystemColumnEncryptionKeyStoreProviders.get(providerName);
-        }
-        return null;
+        return null != globalSystemColumnEncryptionKeyStoreProviders && globalSystemColumnEncryptionKeyStoreProviders
```
Comment (cheenamalhotra) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: Applies here too.. please check this file thoroughly.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 11]
```diff
@@ -6355,7 +6267,7 @@ public void checkConnect() throws SecurityException {
     }
 
     /**
-     * Throws a <code>SecurityException</code> if the calling thread is not allowed to dynamic link the library code.
+     * Checks if the calling thread is allowed to dynamic link the library code.
```
Comment (rene-ye) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: Checks if the calling thread is allowed to dynamically link the library code.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 12]
```diff
@@ -1287,9 +1275,9 @@ Connection connectInternal(Properties propsIn,
                         .getProperty(SQLServerDriverStringProperty.HOSTNAME_IN_CERTIFICATE.toString());
             }
 
+            // if hostNameInCertificate has a legitimate value (and not empty or null),
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: Multi-line.

File: src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java [hunk 13]
```diff
@@ -71,52 +73,30 @@
     private static Boolean isAzureDW = null;
     private static Boolean isAzureMI = null;
 
-    /*
-     * SERVERPROPERTY('EngineEdition') can be used to determine whether the db server is SQL Azure.
-     * It should return 6 for SQL Azure DW. This is more reliable than @@version or
-     * serverproperty('edition').
-     * Reference: http://msdn.microsoft.com/en-us/library/ee336261.aspx
+    /**
+     * Check if connected to Azure server.
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java]: Checks if the connection is established to an Azure server. Please make similar changes to below.

File: src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java [hunk 14]
```diff
@@ -701,4 +681,26 @@ public static boolean supportJDBC43(Connection con) throws SQLException {
...
+     * in development process.
+     *
+     * @return root bundle.
+     */
+    private static ResourceBundle getDefaultLocaleBundle() {
+        return ResourceBundle.getBundle("com.microsoft.sqlserver.jdbc.SQLServerResource", Locale.getDefault());
+    }
+
+    /**
+     * Creates a regex where all '{#}' fields will return true for any value when calling match
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/TestUtils.java]: Add period to the end of this comment for consistency.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 15]
```diff
@@ -4566,24 +4510,18 @@ private void onFeatureExtAck(byte featureId, byte[] data) throws SQLServerExcept
...
-                if (connectionlogger.isLoggable(Level.SEVERE)) {
-                    connectionlogger.severe(toString() + " Unknown feature ack.");
-                }
                 throw new SQLServerException(SQLServerException.getErrString("R_UnknownFeatureAck"), null);
             }
         }
     }
 
-    /*
+    /**
```
Comment (rene-ye) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: Typo?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 16]
```diff
@@ -1744,30 +1736,25 @@ else if (0 == requestedPacketSize)
...
-            xopenStates = booleanPropertyOn(sPropKey, activeConnectionProperties.getProperty(sPropKey));
+            xopenStates = isBooleanPropertyOn(sPropKey, activeConnectionProperties.getProperty(sPropKey));
 
             sPropKey = SQLServerDriverStringProperty.RESPONSE_BUFFERING.toString();
-            responseBuffering = null;
-            if (activeConnectionProperties.getProperty(sPropKey) != null
-                    && activeConnectionProperties.getProperty(sPropKey).length() > 0) {
-                responseBuffering = activeConnectionProperties.getProperty(sPropKey);
-            }
+            responseBuffering = activeConnectionProperties.getProperty(sPropKey) != null
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: This `if` statement was checking for empty string too, now it is only against `null`. 
`activeConnectionProperties.getProperty(sPropKey).length() > 0`

Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: good catch, thx

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 17]
```diff
@@ -4759,10 +4764,10 @@ final boolean complete(LogonCommand logonCommand, TDSReader tdsReader) throws SQ
         int dataLen = 0;
 
         // Denali --> TDS 7.4, Katmai (10.0) & later 7.3B, Prelogin disconnects anything older
-        tdsVersion = serverMajorVersion >= 11 ? TDS.VER_DENALI
-                                              : serverMajorVersion >= 10 ? TDS.VER_KATMAI
-                                                                         : serverMajorVersion >= 9 ? TDS.VER_YUKON
-                                                                                                   : TDS.VER_UNKNOWN;
+        tdsVersion = (serverMajorVersion >= 11) ? TDS.VER_DENALI
+                                                : (serverMajorVersion >= 10) ? TDS.VER_KATMAI
```
Comment (cheenamalhotra) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: Needs precise bracketing for nested ternary operations.

Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: revert cos it looks ugly now

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 18]
```diff
@@ -2032,13 +2020,9 @@ private void login(String primary, String primaryInstanceName, int primaryPortNu
...
         // For non-dbmirroring, non-tnir and non-multisubnetfailover scenarios, full time out would be used as time
         // slice.
-        if (isDBMirroring || useParallel) {
-            timeoutUnitInterval = (long) (TIMEOUTSTEP * timerTimeout);
-        } else if (useTnir) {
-            timeoutUnitInterval = (long) (TIMEOUTSTEP_TNIR * timerTimeout);
-        } else {
-            timeoutUnitInterval = timerTimeout;
-        }
+        timeoutUnitInterval = (isDBMirroring || useParallel) ? (long) (TIMEOUTSTEP * timerTimeout)
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: Why do we need to do this? This impacts readability a lot.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 19]
```diff
@@ -2377,21 +2350,19 @@ ServerPortPlaceHolder primaryPermissionCheck(String primary, String primaryInsta
...
+    /**
+     * Get time remaining to timer expiry
+     * 
+     * @param timerExpire
+     * @return remaining time to expiry
+     */
+    static int timerRemaining(long timerExpire) {
+        long remaining = timerExpire - System.currentTimeMillis();
+        // maximum timeout the socket takes is int max, minimum is at least 1 ms
+        return (int) ((remaining > Integer.MAX_VALUE) ? Integer.MAX_VALUE : (remaining <= 0) ? 1 : remaining);
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: same here as above

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 20]
```diff
@@ -2466,6 +2437,9 @@ void Prelogin(String serverName, int portNumber) throws SQLServerException {
             fedAuthRequiredByUser = true;
         }
 
+        fedAuthRequiredByUser = (!authenticationString.equalsIgnoreCase(SqlAuthentication.NotSpecified.toString()))
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: Why was this line added? it looks like a duplicate of line 2436?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 21]
```diff
@@ -3265,7 +3240,7 @@ final void poolCloseEventNotify() throws SQLServerException {
             // Also if the app closes a connection handle before committing or rolling back the uncompleted
             // transaction may lock other updates/queries so close the transaction now.
             if (!databaseAutoCommitMode && !(pooledConnectionParent instanceof XAConnection)) {
-                connectionCommand("IF @@TRANCOUNT > 0 ROLLBACK TRAN" /* +close connection */, "close connection");
+                connectionCommand("IF @@TRANCOUNT > 0 ROLLBACK TRAN", "close connection");
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: Any reason why we're removing this comment?

Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java]: cos it says "close connection" right after :)
